### DISCUSSION TITLE
Object Creation - Domain Entities

### DISCUSSION THREAD

**Discussion by tomazp** [techexcellenceio/techexcellence#21] [Q&A]
Hello everyone.

I've been wondering about the best way to construct domain entities. I often read that domain entity creation should be delegated to the domain layer via factory - either factory method or dedicated factory class.

But this doesn't seem practical when domain entities are large (many fields) which happens often in the real applications.

So far I've been using a builder for this while keeping the constructors private. But both the Builder and the build method are public. In practice the entity ends up being assembled in a command handler. 

Am I approaching this correctly? Should domain object creation be completely hidden from the outside?

Thank you in advance.

**Comment by CromixPT** [comment]
While I still consider myself a learner and a newbie, I will try to share my understanding and experience.

While you can try to delegate the entity creation and it is often seen in most demos, I also seen quite a few scenarios where you have a public constructor to build the domain entity.

Let's pick banking-kata-dotnet, while having a builder factory for the bank account, that is used in multiple places, we can also see that there is a public constructor: [BankAccount.cs](https://github.com/valentinacupac/banking-kata-dotnet/blob/a7832bc5ee6d7992264530be1d30803cca48abd0/src/Optivem.Kata.Banking.Core/Domain/BankAccounts/BankAccount.cs#L8)

And you can see it being used in the command handler: [OpenAccountUserCase.cs](https://github.com/valentinacupac/banking-kata-dotnet/blob/a7832bc5ee6d7992264530be1d30803cca48abd0/src/Optivem.Kata.Banking.Core/UseCases/OpenAccount/OpenAccountUseCase.cs#L36)

That said, I think there is not a single right way to do so, and in the end it might be related with personal or team preference and coding standards. 

I'm sure some one more experienced will help out clearing the doubt @tomazp
